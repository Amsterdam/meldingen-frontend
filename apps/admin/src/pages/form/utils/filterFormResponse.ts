import type { FormInput } from '@meldingen/api-client'
import {
  FormCheckboxComponentInputSchema,
  FormComponentInputValidateSchema,
  FormInputSchema,
  FormPanelComponentInputSchema,
  FormRadioComponentInputSchema,
  FormSelectComponentInputSchema,
  FormTextAreaComponentInputSchema,
  FormTextFieldComponentInputSchema,
} from '@meldingen/api-client'
import filter from 'uber-json-schema-filter'

const filterBySchemaPerType = (obj: any) => {
  switch (obj.type) {
    case 'panel':
      return filter(FormPanelComponentInputSchema, obj)
    case 'radio':
      return filter(FormRadioComponentInputSchema, obj)
    case 'select':
      return filter(FormSelectComponentInputSchema, obj)
    case 'selectboxes':
      return filter(FormCheckboxComponentInputSchema, obj)
    case 'textarea':
      // Add autoExpand to object if it doesn't exist, the builder doesn't do that by default
      if (!Object.hasOwn(obj, 'autoExpand')) {
        return filter(FormTextAreaComponentInputSchema, { ...obj, autoExpand: false })
      }

      return filter(FormTextAreaComponentInputSchema, obj)
    case 'textfield':
      return filter(FormTextFieldComponentInputSchema, obj)
    default:
      throw Error(`Type ${obj.type} is unknown, please add it to filterFormResponse.`)
  }
}

const getFilteredValidateObject = (validateObj: any) => {
  const validate = filter(FormComponentInputValidateSchema, validateObj)

  // Explicitly remove the 'json' key if its value is an empty string, the API doesn't accept that
  if (validate?.json === '') {
    delete validate.json
  }

  return validate
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const filterFormResponse = (obj: any): FormInput => {
  // This function is used to filter an additional questions form, which is always
  // a wizard with panels containing questions. Therefore the form has a fixed depth of two levels.
  const firstLevelComponents = obj.components.map((firstLevelComponent: any) => {
    const secondLevelComponents = firstLevelComponent.components.map((secondLevelComponent: any) => ({
      ...filterBySchemaPerType(secondLevelComponent),
      validate: getFilteredValidateObject(secondLevelComponent.validate),
    }))

    const filteredObject = {
      ...filterBySchemaPerType(firstLevelComponent),
      components: secondLevelComponents,
      validate: getFilteredValidateObject(firstLevelComponent.validate),
    }

    // Explicitly remove the 'validate' key for panel components, the API doesn't accept that
    if (filteredObject.type === 'panel') {
      delete filteredObject.validate
    }

    return filteredObject
  })

  return {
    ...filter(FormInputSchema, obj),
    components: firstLevelComponents,
  }
}
