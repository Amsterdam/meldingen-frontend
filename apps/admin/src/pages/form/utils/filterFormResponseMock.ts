import type { FormType } from '@formio/react'

const mockPanels = [
  {
    title: 'Page 1',
    label: 'Page 1',
    type: 'panel',
    key: 'page1',
    id: 'er2mpjq',
    input: false,
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: false,
    hidden: false,
    clearOnHide: false,
    refreshOn: '',
    redrawOn: '',
    tableView: false,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    description: '',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: null,
    attributes: {},
    validateOn: 'change',
    validate: {
      required: false,
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
    },
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    tree: false,
    lazyLoad: false,
    theme: 'default',
    breadcrumb: 'default',
  },
  // {
  //   title: 'Page 2',
  //   label: 'Page 2',
  //   type: 'panel',
  //   key: 'page2',
  //   id: 'ehqvlem',
  //   input: false,
  //   placeholder: '',
  //   prefix: '',
  //   customClass: '',
  //   suffix: '',
  //   multiple: false,
  //   defaultValue: null,
  //   protected: false,
  //   unique: false,
  //   persistent: false,
  //   hidden: false,
  //   clearOnHide: false,
  //   refreshOn: '',
  //   redrawOn: '',
  //   tableView: false,
  //   modalEdit: false,
  //   dataGridLabel: false,
  //   labelPosition: 'top',
  //   description: '',
  //   errorLabel: '',
  //   tooltip: '',
  //   hideLabel: false,
  //   tabindex: '',
  //   disabled: false,
  //   autofocus: false,
  //   dbIndex: false,
  //   customDefaultValue: '',
  //   calculateValue: '',
  //   calculateServer: false,
  //   widget: null,
  //   attributes: {},
  //   validateOn: 'change',
  //   validate: {
  //     required: false,
  //     custom: '',
  //     customPrivate: false,
  //     strictDateValidation: false,
  //     multiple: false,
  //     unique: false,
  //   },
  //   conditional: {
  //     show: null,
  //     when: null,
  //     eq: '',
  //   },
  //   overlay: {
  //     style: '',
  //     left: '',
  //     top: '',
  //     width: '',
  //     height: '',
  //   },
  //   allowCalculateOverride: false,
  //   encrypted: false,
  //   showCharCount: false,
  //   showWordCount: false,
  //   properties: {},
  //   allowMultipleMasks: false,
  //   addons: [],
  //   tree: false,
  //   lazyLoad: false,
  //   theme: 'default',
  //   breadcrumb: 'default',
  // },
]

const mockTextFields = [
  {
    label: 'Text Field',
    description: '',
    key: 'textField1',
    type: 'textfield',
    input: true,
    position: 1,
    validate: {
      json: null,
      required: false,
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      minLength: '',
      maxLength: '',
      pattern: '',
    },
    question: 53,
    id: 'emw2bm8',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: {
      type: 'input',
    },
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    mask: false,
    inputType: 'text',
    inputFormat: 'plain',
    inputMask: '',
    displayMask: '',
    spellcheck: true,
    truncateMultipleSpaces: false,
  },
  {
    label: 'Text Field',
    description: 'Description',
    validate: {
      required: true,
      json: {
        '==': [
          {
            var: 'text',
          },
          'Bob',
        ],
      },
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      minLength: '',
      maxLength: '',
      pattern: '',
    },
    input: true,
    key: 'textField2',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: {
      type: 'input',
    },
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    type: 'textfield',
    mask: false,
    inputType: 'text',
    inputFormat: 'plain',
    inputMask: '',
    displayMask: '',
    spellcheck: true,
    truncateMultipleSpaces: false,
    id: 'et6zs29',
  },
]

const mockTextAreas = [
  {
    label: 'Text Area',
    description: '',
    key: 'textArea1',
    type: 'textarea',
    input: true,
    position: 2,
    validate: {
      json: null,
      required: false,
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      minLength: '',
      maxLength: '',
      pattern: '',
      minWords: '',
      maxWords: '',
    },
    autoExpand: false,
    maxCharCount: null,
    question: 54,
    id: 'erhvzak',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: {
      type: 'input',
    },
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    mask: false,
    inputType: 'text',
    inputFormat: 'html',
    inputMask: '',
    displayMask: '',
    spellcheck: true,
    truncateMultipleSpaces: false,
    rows: 3,
    wysiwyg: false,
    editor: '',
    fixedSize: true,
  },
  {
    label: 'Text Area',
    description: 'Description',
    maxCharCount: '80',
    validate: {
      required: true,
      json: {
        '==': [
          {
            var: 'text',
          },
          'Bob',
        ],
      },
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      minLength: '',
      maxLength: '',
      pattern: '',
      minWords: '',
      maxWords: '',
    },
    input: true,
    key: 'textArea2',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: {
      type: 'input',
    },
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    type: 'textarea',
    mask: false,
    inputType: 'text',
    inputFormat: 'html',
    inputMask: '',
    displayMask: '',
    spellcheck: true,
    truncateMultipleSpaces: false,
    rows: 3,
    wysiwyg: false,
    editor: '',
    fixedSize: true,
    id: 'ewrg6od',
    autoExpand: true,
  },
]

const mockSelectBoxes = [
  {
    label: 'Select Boxes',
    description: '',
    key: 'selectBoxes',
    type: 'selectboxes',
    input: true,
    position: 3,
    validate: {
      json: null,
      required: false,
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    values: [
      {
        label: 'One',
        value: 'one',
        position: 1,
      },
      {
        label: 'Two',
        value: 'two',
        position: 2,
      },
    ],
    question: 55,
    id: 'e7rvihg',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: false,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: null,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    inputType: 'checkbox',
    data: {
      url: '',
    },
    fieldSet: false,
    inline: false,
  },
  {
    label: 'Select Boxes',
    description: 'Description',
    validate: {
      required: true,
      json: {
        '==': [
          {
            var: 'text',
          },
          'Bob',
        ],
      },
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    values: [
      {
        label: 'One',
        value: 'one',
      },
      {
        label: 'Two',
        value: 'two',
      },
    ],
    input: true,
    key: 'selectBoxes1',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: false,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: null,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    type: 'selectboxes',
    inputType: 'checkbox',
    data: {
      url: '',
    },
    fieldSet: false,
    inline: false,
    id: 'e577fo9',
  },
]

const mockSelects = [
  {
    label: 'Select',
    widget: 'html5',
    placeholder: '',
    description: '',
    validate: {
      required: false,
      json: '',
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    data: {
      values: [
        {
          label: 'One',
          value: 'one',
        },
        {
          label: 'Two',
          value: 'two',
        },
      ],
      json: '',
      url: '',
      resource: '',
      custom: '',
    },
    input: true,
    key: 'select',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    type: 'select',
    idPath: 'id',
    clearOnRefresh: false,
    limit: 100,
    valueProperty: '',
    lazyLoad: true,
    filter: '',
    searchEnabled: true,
    searchDebounce: 0.3,
    searchField: '',
    minSearch: 0,
    readOnlyValue: false,
    selectFields: '',
    selectThreshold: 0.3,
    uniqueOptions: false,
    fuseOptions: {
      include: 'score',
      threshold: 0.3,
    },
    indexeddb: {
      filter: {},
    },
    customOptions: {},
    useExactSearch: false,
    id: 'e33dqt',
  },
  {
    label: 'Select',
    widget: 'html5',
    placeholder: 'Placeholder',
    description: 'Description',
    validate: {
      required: true,
      json: {
        '==': [
          {
            var: 'text',
          },
          'Bob',
        ],
      },
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    data: {
      values: [
        {
          label: 'One',
          value: 'one',
        },
        {
          label: 'Two',
          value: 'two',
        },
      ],
      json: '',
      url: '',
      resource: '',
      custom: '',
    },
    input: true,
    key: 'select1',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: true,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    type: 'select',
    idPath: 'id',
    clearOnRefresh: false,
    limit: 100,
    valueProperty: '',
    lazyLoad: true,
    filter: '',
    searchEnabled: true,
    searchDebounce: 0.3,
    searchField: '',
    minSearch: 0,
    readOnlyValue: false,
    selectFields: '',
    selectThreshold: 0.3,
    uniqueOptions: false,
    fuseOptions: {
      include: 'score',
      threshold: 0.3,
    },
    indexeddb: {
      filter: {},
    },
    customOptions: {},
    useExactSearch: false,
    id: 'e4p4x3ea',
  },
]

const mockRadios = [
  {
    label: 'Radio',
    description: '',
    validate: {
      required: false,
      json: '',
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    values: [
      {
        label: 'One',
        value: 'one',
      },
      {
        label: 'Two',
        value: 'two',
      },
    ],
    input: true,
    key: 'radio',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: false,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: null,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    type: 'radio',
    inputType: 'radio',
    data: {
      url: '',
    },
    fieldSet: false,
    id: 'e44f4c',
  },
  {
    label: 'Radio',
    description: 'Description',
    validate: {
      required: true,
      json: {
        '==': [
          {
            var: 'text',
          },
          'Bob',
        ],
      },
      custom: '',
      customPrivate: false,
      strictDateValidation: false,
      multiple: false,
      unique: false,
      onlyAvailableItems: false,
    },
    values: [
      {
        label: 'One',
        value: 'one',
      },
      {
        label: 'Two',
        value: 'two',
      },
    ],
    input: true,
    key: 'radio1',
    placeholder: '',
    prefix: '',
    customClass: '',
    suffix: '',
    multiple: false,
    defaultValue: null,
    protected: false,
    unique: false,
    persistent: true,
    hidden: false,
    clearOnHide: true,
    refreshOn: '',
    redrawOn: '',
    tableView: false,
    modalEdit: false,
    dataGridLabel: false,
    labelPosition: 'top',
    errorLabel: '',
    tooltip: '',
    hideLabel: false,
    tabindex: '',
    disabled: false,
    autofocus: false,
    dbIndex: false,
    customDefaultValue: '',
    calculateValue: '',
    calculateServer: false,
    widget: null,
    attributes: {},
    validateOn: 'change',
    conditional: {
      show: null,
      when: null,
      eq: '',
    },
    overlay: {
      style: '',
      left: '',
      top: '',
      width: '',
      height: '',
    },
    allowCalculateOverride: false,
    encrypted: false,
    showCharCount: false,
    showWordCount: false,
    properties: {},
    allowMultipleMasks: false,
    addons: [],
    dataSrc: 'values',
    authenticate: false,
    ignoreCache: false,
    template: '<span>{{ item.label }}</span>',
    type: 'radio',
    inputType: 'radio',
    data: {
      url: '',
    },
    fieldSet: false,
    id: 'ej0tb08',
  },
]

export const mockData: FormType & { id: number; classification: string | null } = {
  display: 'wizard',
  title: 'Test title',
  id: 2,
  classification: null,
  components: [
    {
      ...mockPanels[0],
      components: [mockTextFields[0], mockTextAreas[0], mockSelectBoxes[0], mockSelects[0], mockRadios[0]],
    },
    {
      ...mockPanels[1],
      components: [mockTextFields[1], mockTextAreas[1], mockSelectBoxes[1], mockSelects[1], mockRadios[1]],
    },
  ],
}
