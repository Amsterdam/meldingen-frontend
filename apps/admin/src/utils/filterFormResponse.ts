const allowLists: Record<string, string[]> = {
  panel: ['label', 'key', 'type', 'input', 'components'],
  textarea: ['label', 'description', 'key', 'type', 'input', 'autoExpand', 'showCharCount'],
}

// TODO: second 'any' should be a type supplied by the BE?
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const filterFormResponse = (obj: any): any => {
  // If the input is a JS primitive, return the unmodified input
  if (obj !== Object(obj)) {
    return obj
  }

  // If the input is an array, run this function recursively on all items of the array
  if (Array.isArray(obj)) {
    return obj.map((item) => filterFormResponse(item))
  }

  return (
    Object.keys(obj)
      // If input has a key called 'type', and the value of 'type' is present in the keys of the allowLists object,
      // filter keys based on that specific allow list. Otherwise, all keys are passed.
      .filter((key) => {
        if (Object.keys(allowLists).includes(obj?.type)) {
          return allowLists[obj.type].includes(key)
        }

        return true
      })
      // Reduce the input by recursively calling the function on each value and assigning it to the key
      .reduce((acc, x) => Object.assign(acc, { [x]: filterFormResponse(obj[x]) }), {})
  )
}
