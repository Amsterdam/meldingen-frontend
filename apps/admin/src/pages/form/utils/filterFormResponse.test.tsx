import { filterFormResponse } from './filterFormResponse'
import { mockData } from './filterFormResponseMock'

// This test compares the output of the function to a snapshot.
// To update the snapshot, run the following command in apps/admin:
// npx vitest -u
describe('filterFormResponse', () => {
  it('renders the correct output', () => {
    expect(filterFormResponse(mockData)).toMatchSnapshot()
  })
})
