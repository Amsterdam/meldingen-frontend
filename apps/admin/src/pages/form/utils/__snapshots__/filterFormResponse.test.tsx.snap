// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filterFormResponse > renders the correct output 1`] = `
{
  "classification": null,
  "components": [
    {
      "components": [
        {
          "description": "",
          "input": true,
          "key": "textField1",
          "label": "Text Field",
          "type": "textfield",
          "validate": {
            "json": null,
            "required": false,
          },
        },
        {
          "autoExpand": false,
          "description": "",
          "input": true,
          "key": "textArea1",
          "label": "Text Area",
          "maxCharCount": null,
          "type": "textarea",
          "validate": {
            "json": null,
            "required": false,
          },
        },
        {
          "description": "",
          "input": true,
          "key": "selectBoxes",
          "label": "Select Boxes",
          "type": "selectboxes",
          "validate": {
            "json": null,
            "required": false,
          },
          "values": [
            {
              "label": "One",
              "position": 1,
              "value": "one",
            },
            {
              "label": "Two",
              "position": 2,
              "value": "two",
            },
          ],
        },
        {
          "data": {
            "custom": "",
            "json": "",
            "resource": "",
            "url": "",
            "values": [
              {
                "label": "One",
                "value": "one",
              },
              {
                "label": "Two",
                "value": "two",
              },
            ],
          },
          "description": "",
          "input": true,
          "key": "select",
          "label": "Select",
          "placeholder": "",
          "type": "select",
          "validate": {
            "required": false,
          },
          "widget": "html5",
        },
        {
          "description": "",
          "input": true,
          "key": "radio",
          "label": "Radio",
          "type": "radio",
          "validate": {
            "required": false,
          },
          "values": [
            {
              "label": "One",
              "value": "one",
            },
            {
              "label": "Two",
              "value": "two",
            },
          ],
        },
      ],
      "input": false,
      "key": "page1",
      "label": "Page 1",
      "title": "Page 1",
      "type": "panel",
    },
    {
      "components": [
        {
          "description": "Description",
          "input": true,
          "key": "textField2",
          "label": "Text Field",
          "type": "textfield",
          "validate": {
            "json": {
              "==": [
                {
                  "var": "text",
                },
                "Bob",
              ],
            },
            "required": true,
          },
        },
        {
          "autoExpand": true,
          "description": "Description",
          "input": true,
          "key": "textArea2",
          "label": "Text Area",
          "maxCharCount": "80",
          "type": "textarea",
          "validate": {
            "json": {
              "==": [
                {
                  "var": "text",
                },
                "Bob",
              ],
            },
            "required": true,
          },
        },
        {
          "description": "Description",
          "input": true,
          "key": "selectBoxes1",
          "label": "Select Boxes",
          "type": "selectboxes",
          "validate": {
            "json": {
              "==": [
                {
                  "var": "text",
                },
                "Bob",
              ],
            },
            "required": true,
          },
          "values": [
            {
              "label": "One",
              "value": "one",
            },
            {
              "label": "Two",
              "value": "two",
            },
          ],
        },
        {
          "data": {
            "custom": "",
            "json": "",
            "resource": "",
            "url": "",
            "values": [
              {
                "label": "One",
                "value": "one",
              },
              {
                "label": "Two",
                "value": "two",
              },
            ],
          },
          "description": "Description",
          "input": true,
          "key": "select1",
          "label": "Select",
          "placeholder": "Placeholder",
          "type": "select",
          "validate": {
            "json": {
              "==": [
                {
                  "var": "text",
                },
                "Bob",
              ],
            },
            "required": true,
          },
          "widget": "html5",
        },
        {
          "description": "Description",
          "input": true,
          "key": "radio1",
          "label": "Radio",
          "type": "radio",
          "validate": {
            "json": {
              "==": [
                {
                  "var": "text",
                },
                "Bob",
              ],
            },
            "required": true,
          },
          "values": [
            {
              "label": "One",
              "value": "one",
            },
            {
              "label": "Two",
              "value": "two",
            },
          ],
        },
      ],
      "input": false,
      "key": "page2",
      "label": "Page 2",
      "title": "Page 2",
      "type": "panel",
    },
  ],
  "display": "wizard",
  "title": "Test title",
}
`;
